compiler/optimizations/gpuTransforms.cpp

```diff
@@ -1368,16 +1399,36 @@ static VarSymbol* generateNumThreads(BlockStmt* gpuLaunchBlock,
   return numThreads;
 }
 
-static CallExpr* generateGPUCall(GpuKernel& info, VarSymbol* numThreads) {
+static CallExpr* generateGPUCall(GpuKernel& info, BlockStmt* gpuBlock, VarSymbol* numThreads) {
   CallExpr *call = new CallExpr(PRIM_GPU_KERNEL_LAUNCH_FLAT);
 
   call->insertAtTail(info.fn());
 
   call->insertAtTail(numThreads);  // total number of GPU threads
 
-  if (info.blockSize()) {
+  if (auto blockSizeCall = info.blockSizeCall()) {
     // sets blockSize if specified with by "gpu set BlockSize" primitive
-    call->insertAtTail(info.blockSize()->copy());
+    debuggerBreakHere();
+
+    // If the parent of the call is 'gpu primitives' block, it was constructed
+    // specifically to 'fence off' GPU primitives. There may be
+    // some additional temps etc. that are used for computing block size.
+    // We need to lift them, too. Since we're "consuming" the GPU loop,
+    // just modify the parent block in place.
+    if (auto callParent = toBlockStmt(blockSizeCall->parentExpr)) {
+      if (callParent->isGpuPrimitivesBlock()) {
+        for_alist(expr, callParent->body) {
+          if (isCallToPrimitiveWeShouldNotCopyIntoKernel(toCallExpr(expr))) {
+            expr->remove();
+          }
+        }
```

Comment:
I am not asking you to make this change here, but ideally we should have consumed these blocks/expressions earlier during kernel generation, right? The current implementation of gpuTransforms mixes up analysis and transformation too much for my taste and that's one of my gripes about our compiler implementation. I might start refactoring the implementation to make the separation a bit clear, and if I do that this part is more of an "analysis" than "transformation" to me. What do you think?

Reply:
I think this is a good idea. It's definitely odd to perform transformations while doing analysis, and vice versa.